### PR TITLE
ISSUE-118 PGP: CassandraEncryptedEmailContentStore

### PR DESCRIPTION
**Pull request by vttranlina**
Ref https://github.com/linagora/tmail-backend/issues/118

### CODE REVIEW

File: tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/table/CassandraEncryptedEmailStoreModule.scala [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    .comment("Holds content of encrypted email")
+    .options((options: Options) => options
+      .caching(ALL, rows(DEFAULT_CACHED_ROW_PER_PARTITION))
+      .compressionOptions(SchemaBuilder.lz4.withChunkLengthInKb(8)))
+    .statement((statement: Create) => statement
+      .addPartitionKey(EncryptedEmailTable.MESSAGE_ID, uuid)
+      .addColumn(EncryptedEmailTable.ENCRYPTED_PREVIEW, text())
+      .addColumn(EncryptedEmailTable.ENCRYPTED_HTML, text())
+      .addColumn(EncryptedEmailTable.HAS_ATTACHMENT, cboolean())
+      .addColumn(EncryptedEmailTable.ENCRYPTED_ATTACHMENT_METADATA, text()))
```
Comment (chibenwa) [inline tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/table/CassandraEncryptedEmailStoreModule.scala:22]: Likeky `frozen(list(text()))` as a type here:

 - We get several attachement per entry here
 - Frozen as once inserted we will not modify the collection.

File: tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/table/CassandraEncryptedEmailStoreModule.scala [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+import org.apache.james.backends.cassandra.utils.CassandraConstants.DEFAULT_CACHED_ROW_PER_PARTITION
+
+object CassandraEncryptedEmailStoreModule {
+  val EMAIL_CONTENT_MODULE: CassandraModule = CassandraModule.table(EncryptedEmailTable.TABLE_NAME)
+    .comment("Holds content of encrypted email")
+    .options((options: Options) => options
+      .caching(ALL, rows(DEFAULT_CACHED_ROW_PER_PARTITION))
+      .compressionOptions(SchemaBuilder.lz4.withChunkLengthInKb(8)))
+    .statement((statement: Create) => statement
+      .addPartitionKey(EncryptedEmailTable.MESSAGE_ID, uuid)
```
Comment (chibenwa) [inline tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/table/CassandraEncryptedEmailStoreModule.scala:18]: ... timeUuid maybe?

File: tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/table/CassandraEncryptedEmailStoreModule.scala [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+  val MESSAGE_ID_BLOB_ID_MODULE: CassandraModule = CassandraModule.table(EncryptedEmailIdBlobIdTable.TABLE_NAME)
+    .comment("Holds mapping among messageId, blobId and position")
+    .options((options: Options) => options
+      .caching(ALL, rows(DEFAULT_CACHED_ROW_PER_PARTITION))
+      .compressionOptions(SchemaBuilder.lz4.withChunkLengthInKb(8)))
+    .statement((statement: Create) => statement
+      .addPartitionKey(EncryptedEmailIdBlobIdTable.MESSAGE_ID, uuid)
+      .addColumn(EncryptedEmailIdBlobIdTable.BLOB_ID, text())
+      .addColumn(EncryptedEmailIdBlobIdTable.POSITION, cint()))
```
Comment (chibenwa) [inline tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/table/CassandraEncryptedEmailStoreModule.scala:33]: What is this `Position` ?

Comment (vttranlina) [reply tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/table/CassandraEncryptedEmailStoreModule.scala:33]: for `def retrieveAttachmentContent(messageId: MessageId, position: Int): BlobId` method

File: tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/table/CassandraEncryptedEmailStoreModule.scala (hunk 1 above)
Comment (vttranlina) [reply tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/table/CassandraEncryptedEmailStoreModule.scala:22]: Nice freature, Thanks

File: tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/table/CassandraEncryptedEmailStoreModule.scala (hunk 3 above)
Comment (chibenwa) [reply tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/table/CassandraEncryptedEmailStoreModule.scala:33]: Then it should be a clustering column if you wish to use it in select collumns.

BTW using a frozen UDT list you can put it in the table above I guess...

File: tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/table/CassandraEncryptedEmailStoreModule.scala (hunk 1 above)
Comment (vttranlina) [reply tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/table/CassandraEncryptedEmailStoreModule.scala:22]: `ENCRYPTED_ATTACHMENT_METADATA` is String (not List[String]), So I think we don't need `frozenList` here

Comment (chibenwa) [reply tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/table/CassandraEncryptedEmailStoreModule.scala:22]: Yes I remember now it is an encrypted JSON

File: tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/table/CassandraEncryptedEmailStoreModule.scala (hunk 2 above)
Comment (vttranlina) [reply tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/table/CassandraEncryptedEmailStoreModule.scala:18]: Maybe is `text()`, This column will storage `messageId.serialize()`

File: tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/table/CassandraEncryptedEmailContentStore.scala [hunk 4]
```diff
@@ -17,25 +23,43 @@ case class EncryptedEmailContentStoreCassandraModule() extends AbstractModule {
...
+      .map(encryptedEmailDetailedView => EncryptedEmailFastView.from(encryptedEmailDetailedView))
 
-  override def retrieveFastView(messageId: MessageId): EncryptedEmailFastView = ???
+  override def retrieveDetailedView(messageId: MessageId): Publisher[EncryptedEmailDetailedView] =
+    encryptedEmailDAO.get(messageId)
 
-  override def retrieveDetailedView(messageId: MessageId): EncryptedEmailDetailedView = ???
+  override def retrieveAttachmentContent(messageId: MessageId, position: Int): Publisher[BlobId] =
+    encryptedEmailDAO.getBlobId(messageId, position)
+      .onErrorResume((error)=>throw AttachmentNotFoundException(messageId, position))
```
Comment (vttranlina) [inline tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/table/CassandraEncryptedEmailContentStore.scala:60]: I have a reactor problem here
When run test
```
Expecting actual throwable to be an instance of:
  com.linagora.tmail.encrypted.AttachmentNotFoundException
but was:
  reactor.core.Exceptions$ReactiveException: com.linagora.tmail.encrypted.AttachmentNotFoundException
```

File: tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/table/CassandraEncryptedEmailStoreModule.scala (hunk 3 above)
Comment (vttranlina) [reply tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/table/CassandraEncryptedEmailStoreModule.scala]: I added a new column in the above table. 
```
.addColumn(EncryptedEmailTable.POSITION_BLOB_ID_MAPPING, frozenMap(cint(), text())))
```
So this table is unnecessary => removed

File: tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/table/CassandraEncryptedEmailStoreModule.scala (hunk 2 above)
Comment (chibenwa) [reply tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/table/CassandraEncryptedEmailStoreModule.scala]: Wh do we need to serialie it? Can't we assume this is a CassandraMessageId?

File: tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/table/CassandraEncryptedEmailContentStore.scala (hunk 4 above)
Comment (chibenwa) [reply tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/table/CassandraEncryptedEmailContentStore.scala:60]: Hmm not really...

You can switch the assertion in the tests to hasCause for instance.

BTW you can use onErrorMap to aoid this uggy throw I imagine...

Comment (vttranlina) [reply tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/table/CassandraEncryptedEmailContentStore.scala:60]: I fixed it,  `AttachmentNotFoundException extends Exception`  => `AttachmentNotFoundException extends RuntimeException`.

File: tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/table/CassandraEncryptedEmailStoreModule.scala [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+import com.datastax.driver.core.schemabuilder.{Create, SchemaBuilder}
+import org.apache.james.backends.cassandra.components.CassandraModule
+import org.apache.james.backends.cassandra.utils.CassandraConstants.DEFAULT_CACHED_ROW_PER_PARTITION
+
+object CassandraEncryptedEmailStoreModule {
+  val MODULE: CassandraModule = CassandraModule.table(EncryptedEmailTable.TABLE_NAME)
+    .comment("Holds content of encrypted email")
+    .options((options: Options) => options
+      .caching(ALL, rows(DEFAULT_CACHED_ROW_PER_PARTITION))
+      .compressionOptions(SchemaBuilder.lz4.withChunkLengthInKb(8)))
```
Comment (chibenwa) [inline tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/table/CassandraEncryptedEmailStoreModule.scala:16]: I'm not sure configuring compression make sense here (encrypted content can likely be large...)

File: tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/table/CassandraEncryptedEmailStoreModule.scala [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+import com.datastax.driver.core.schemabuilder.Create.Options
+import com.datastax.driver.core.schemabuilder.SchemaBuilder.KeyCaching.ALL
+import com.datastax.driver.core.schemabuilder.SchemaBuilder.rows
+import com.datastax.driver.core.schemabuilder.{Create, SchemaBuilder}
+import org.apache.james.backends.cassandra.components.CassandraModule
+import org.apache.james.backends.cassandra.utils.CassandraConstants.DEFAULT_CACHED_ROW_PER_PARTITION
+
+object CassandraEncryptedEmailStoreModule {
+  val MODULE: CassandraModule = CassandraModule.table(EncryptedEmailTable.TABLE_NAME)
+    .comment("Holds content of encrypted email")
```
Comment (chibenwa) [inline tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/table/CassandraEncryptedEmailStoreModule.scala:13]: in order to ease JMAP display.

**Review by chibenwa — COMMENTED** [review]
Looks good!

File: tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/table/CassandraEncryptedEmailStoreModule.scala (hunk 5 above)
Comment (vttranlina) [reply tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/table/CassandraEncryptedEmailStoreModule.scala:16]: Honestly, I am not sure too. Maybe I will spend time digging into these options soon.

Comment (chibenwa) [reply tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/table/CassandraEncryptedEmailStoreModule.scala:16]: If we are not sure please let the defaults here: we can tune it later ;-)

**Review by Arsnael — COMMENTED** [review]
Read it. Nothing to add :)

**Review by Arsnael — APPROVED** [review]

File: tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/table/CassandraEncryptedEmailStoreModule.scala [hunk 7]
```diff
@@ -10,10 +10,9 @@ import org.apache.james.backends.cassandra.utils.CassandraConstants.DEFAULT_CACH
 
 object CassandraEncryptedEmailStoreModule {
   val MODULE: CassandraModule = CassandraModule.table(EncryptedEmailTable.TABLE_NAME)
-    .comment("Holds content of encrypted email")
+    .comment("In order to ease the JMAP display")
```
Comment (chibenwa) [inline tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/table/CassandraEncryptedEmailStoreModule.scala:13]:
```suggestion
    .comment("olds content of encrypted email in order to ease the JMAP display")
```

Comment (vttranlina) [reply tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/table/CassandraEncryptedEmailStoreModule.scala:13]: okay, Miss understand :))

**Comment by vttranlina** [comment]
Waiting https://github.com/linagora/tmail-backend/pull/177 merge for resolve conflict

**Comment by vttranlina** [comment]
Rebase master
Squash fixup commit
Fix comment: https://github.com/linagora/tmail-backend/pull/178#discussion_r661064819

File: tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/table/CassandraEncryptedEmailContentStore.scala [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    Multibinder.newSetBinder(binder, classOf[CassandraModule])
+      .addBinding()
+      .toInstance(CassandraEncryptedEmailStoreModule.MODULE)
+  }
+}
+
+class CassandraEncryptedEmailContentStore @Inject()(blobStore: BlobStore,
+                                                    bucketName: BucketName,
+                                                    storagePolicy: StoragePolicy,
+                                                    encryptedEmailDAO: CassandraEncryptedEmailDAO) extends EncryptedEmailContentStore {
```
Comment (chibenwa) [inline tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/table/CassandraEncryptedEmailContentStore.scala:33]: Storage policy is always low cost. Use blobStore.getDefaultBucket.

**Review by chibenwa — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Pull request merged**